### PR TITLE
Refactor the ElasticsearchClient integration tests

### LINKED ISSUE #9320: Refactor ElasticsearchClient, splitting into separate concerns for easier testing
**Description**

The `ElasticsearchClient` is the one class in the exporter which communicates with the actual Elastic instance. Currently, it's responsibilities are:

- Configures and manage the Elastic REST client
- Network communication: (de)serailization, response/error handling, etc.
- Read templates from the resources, parse them, and substitute certain configuration options
- Compute the index name, document ID, etc., properties of each document.
- Serialization of records and templates
- Buffer serialized indexing commands in memory
- Configurable logic for when flushing should occur

I would propose to do the following:

- [x] #9327
- [x] #9328
- [x] #9329
- [x] #9340
- [ ] #9330
- [x] #9331

The client will now only deal with network communication and in memory buffering of the documents. I would recommend moving the flushing/buffering out into its own component as well later, but the above is what provides the biggest win, I think, in terms of maintainability and test-ability.

The last two steps will be blocked by the first three, I think, but the first three can be done in parallel. Finally, refactoring the tests may include adding new testing utilities, but these will later be reused in #9321.

### LINKED ISSUE #9331: Refactor client's integration tests

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR refactors the `ElasticsearchClient` integration tests. This somewhat increases test coverage by adding new tests with a reduced blast radius, as opposed to testing the client indirectly via the exporter running in a broker.

## Related issues

closes #9331
closes #9320 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Review by remcowesterhoud — COMMENTED** [review]
@npepinpe These changes look good 🏆 I won't approve it yet as I don't want this to be merged just yet, since the other PR's are merging into this branch.

**Review by remcowesterhoud — APPROVED** [review]
I realised I can just approve it. We can always target the other PRs to main 🤷 

Please have a look at my question and minor comment.

File: exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/TestSupport.java [hunk 1]
```diff
@@ -10,11 +10,33 @@
...
+  private static final DockerImageName ELASTIC_IMAGE =
+      DockerImageName.parse("docker.elastic.co/elasticsearch/elasticsearch")
+          .withTag(RestClient.class.getPackage().getImplementationVersion());
+
   private TestSupport() {}
 
+  /**
+   * Returns an Elasticsearch container pointing at the same version as the {@link RestClient}.
+   *
+   * <p>The container is configured to use 750m of heap and 750m of direct memory. This is required
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/TestSupport.java:28]: You're right, this is a left over from copy pasting :see_no_evil:

File: exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchClientIT.java [hunk 2]
```diff
@@ -7,31 +7,77 @@
...
+
+    testClient = new TestClient(config, indexRouter);
+    client =
+        new ElasticsearchClient(
+            config,
+            bulkRequest,
+            RestClientFactory.of(config),
+            indexRouter,
+            templateReader,
+            new ElasticsearchMetrics(1));
```
[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchClientIT.java:71]: It does make sense. I'm stuck on finding a proper name to be honest. It seems to do more than just verification, as it is also used to add users in one of the testcases.

[the remark this answers is not in the document]
Comment (npepinpe) [reply exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchClientIT.java:71]: If that's OK, I'll leave it as is for now. We can always rename it later.

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**